crates/kubelet/src/state/common/mod.rs

```diff
@@ -0,0 +1,106 @@
+//! Contains a set of generic states where there is little provider-specific
+//! logic, and the machinery to use them. This removes the need to write these
+//! states in many providers; instead, the provider need only implement the
+//! GenericProviderState and GenericPodState traits for its state types.
+
+use crate::state::State;
+
+use std::collections::HashMap;
+
+pub mod crash_loop_backoff;
+pub mod error;
+pub mod image_pull;
+pub mod image_pull_backoff;
+pub mod registered;
+pub mod terminated;
+pub mod volume_mount;
+
+/// Types of error condition whose backoff should be tracked independently.
+pub enum BackoffSequence {
+    /// Backoff from a failed image pull.
+    ImagePull,
+    /// Backoff from a pod crash.
+    CrashLoop,
+}
+
+/// Indicates whether a threshold has been triggered.
+pub enum ThresholdTrigger {
+    /// The threshold has been triggered.
+    Triggered,
+    /// The threshold has not been triggered.
+    Untriggered,
+}
+
+/// Exposes provider-wide state in a way that can be consumed by
+/// the generic states.
+#[async_trait::async_trait]
+pub trait GenericProviderState: 'static + Send + Sync {
+    /// Gets a Kubernetes client. This is a provider function to enable the
+    /// provider to control the client configuration as desired.
+    fn client(&self) -> kube::Client;
+    /// Gets the `Store` used by the provider.
+    fn store(&self) -> std::sync::Arc<dyn crate::store::Store + Sync + Send>;
+    /// Gets the path at which to construct temporary directories for volumes.
+    fn volume_path(&self) -> std::path::PathBuf;
+    /// Stops the specified pod. This typically involves tearing down a
+    /// runtime or other execution environment.
+    async fn stop(&self, pod: &crate::pod::Pod) -> anyhow::Result<()>;
+}
+
+/// Exposes pod state in a way that can be consumed by
+/// the generic states.
+#[async_trait::async_trait]
+pub trait GenericPodState: 'static + Send + Sync {
+    /// Stores the pod module binaries for future execution. Typically your
+    /// implementation can just move the modules map into a member field.
+    fn set_modules(&mut self, modules: HashMap<String, Vec<u8>>);
```

Comment:
Is this too WASM specific?

Reply:
Possibly, but it's consistent with the terminology elsewhere (e.g. `fetch_pod_modules`).  If we have a more generic term than 'modules' then I'm happy to use it.